src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,451 @@
+                  same index, and whether the index is unmodified in
+                  the code region.
+        '''
+        array_indices = None
+        scalarizable = True
+        for access in var_accesses[signature].all_accesses:
+            if array_indices is None:
+                array_indices = access.component_indices
+            # For some reason using == on the component_lists doesn't work
+            # so we use [:] notation.
+            elif array_indices[:] != access.component_indices[:]:
+                scalarizable = False
+                break
+            # For each index, we need to check they're not written to in
+            # the loop.
+            flattened_indices = list(itertools.chain.from_iterable(
+                    array_indices))
+            for index in flattened_indices:
+                # Index may not be a Reference, so we need to loop over the
+                # References
+                for ref in index.walk(Reference):
+                    # This Reference could be the symbol for a Call or
+                    # IntrinsicCall, which we don't allow to scalarize
+                    if isinstance(ref.symbol, RoutineSymbol):
+                        scalarizable = False
+                        break
+                    sig, _ = ref.get_signature_and_indices()
+                    if var_accesses[sig].is_written():
+                        scalarizable = False
+                        break
+
+        return scalarizable
+
+    @staticmethod
+    def _check_first_access_is_write(signature: Signature,
+                                     loop: Loop,
+                                     var_accesses: VariablesAccessInfo) \
+            -> bool:
+        '''
+        :param signature: The signature to check.
+        :param loop: The Loop object being transformed.
+        :param var_accesses: The VariableAccessesInfo object containing
+                             signature.
+        :returns: whether the first access to signature is a write.
+        '''
+        if not var_accesses[signature].is_written_first():
+            return False
+        # Need to find the first access and check if its in a conditional.
+        accesses = var_accesses[signature].all_accesses
+        first_node = accesses[0].node
+        ifblock = first_node.ancestor(IfBlock)
+        # If the depth of the ifblock is larger than loop then the write
+        # is in a conditional
+        if ifblock and ifblock.depth > loop.depth:
+            return False
+        return True
+
+    @staticmethod
+    def _get_index_values_from_indices(
+            node: ArrayMixin, indices: List[Node]) -> Tuple[bool, List[Node]]:
+        '''
+        Compute a list of index values for a given node. Looks at loop bounds
+        and range declarations to attempt to convert loop variables to an
+        explicit range, i.e. an access like
+        >>> do i = 1, 100
```

Comment:
The ">>>" makes `python -m doctest src/psyclone/psyir/transformations/scalarization_trans.py` fail. If you want a fortran snippet you can use the rst notation `.. code-block:: fortran` and it will appear properly rendered in the generated documentation

Reply:
I think I've fixed that now.